[clang] Non-object types are non-trivially relocatable

AMP999 opened the pull request:
Both active C++ proposals (P1144 and P2786) agree that `is_trivially_relocatable_v<int&>` and `is_trivially_relocatable_v<int()>` should be false, not true. Only complete object types can be trivially relocatable.

Fixes #67498

cor3ntin commented:
an you add tests  / release notes? (the change itself looks fine to me)

AMP999 commented:
@cor3ntin Are these tests what you had in mind? Are they sufficient? What else should I add?

AMP999 commented:
@cor3ntin Gentle ping!

cor3ntin commented:
@AMP999 Sorry, I'm in a committee meeting this week. tests look good. it's still missing a release note though

AMP999 commented:
Thank you! No problem! I'll add the release notes.

cor3ntin commented on clang/docs/ReleaseNotes.rst:
```diff
@@ -477,10 +477,10 @@ Bug Fixes in This Version
-- Fix crash in __builtin_strncmp and related builtins when the size value
-  exceeded the maximum value representable by int64_t. Fixes
+- Fix crash in ``__builtin_strncmp`` and related builtins when the size value
+  exceeded the maximum value representable by ``int64_t``. Fixes
   (`#64876 <https://github.com/llvm/llvm-project/issues/64876>`_)
-- Fixed an assertion if a function has cleanups and fatal erors.
+- Fixed an assertion if a function has cleanups and fatal errors.
```
Can you revert that and commit it separately in a separate NFC patch? thanks

AMP999 replied:
Sure, it's done!

AMP999 committed:
[clang] Non-object types are non-trivially relocatable

Both active C++ proposals (P1144 and P2786) agree that
`is_trivially_relocatable_v<int&>` and `is_trivially_relocatable_v<int()>`
should be false, not true. Only complete object types
can be trivially relocatable.

Fixes #67498

cor3ntin reviewed (approved):
LGTM, thanks!

AMP999 commented:
> LGTM, thanks!

You're welcome!

AMP999 commented:
@cor3ntin Could you land this for me, please?

cor3ntin commented:
@AMP999 can you resolve the conflict?

AMP999 commented:
@cor3ntin I've fixed that.
